polkadot-protocol/basics/data-encoding.md

```diff
@@ -1,4 +1,110 @@
 ---
 title: Data Encoding
-description: TODO
----
\ No newline at end of file
+description: SCALE codec enables fast, efficient data encoding, ideal for resource-constrained environments like Wasm, supporting custom types and compact encoding.
+---
+
+# Data Encoding
+
+The Polkadot SDK uses a lightweight and efficient encoding/decoding mechanism to optimize data transmission across the network. This mechanism, known as the **SCALE** codec, is used for serializing and deserializing data.
+
+The **SCALE** codec enables communication between the runtime and the outer node. This mechanism is designed for high-performance, copy-free data encoding and decoding in resource-constrained environments like the Polkadot SDK [Wasm runtime](TODO:update-path){target=\_blank}.
+
+It is not self-describing, meaning the decoding context must have full knowledge of the encoded data types. 
+
+Parity's libraries utilize the [`parity-scale-codec`](https://github.com/paritytech/parity-scale-codec){target=\_blank} crate (a Rust implementation of the SCALE codec) to handle encoding and decoding for interactions between RPCs and the runtime.
+
+The `codec` mechanism is ideal for Polkadot SDK-based chains because:
+
+- It is lightweight compared to generic serialization frameworks like [`serde`](https://serde.rs/){target=\_blank}, which add unnecessary bulk to binaries
+- It doesn’t rely on Rust’s `libstd`, making it compatible with `no_std` environments like Wasm runtime
+- It integrates seamlessly with Rust, allowing easy derivation of encoding and decoding logic for new types using `#[derive(Encode, Decode)]`
+
+Defining a custom encoding scheme in the Polkadot SDK-based chains, rather than using an existing Rust codec library, is crucial for enabling cross-platform and multi-language support. 
+
+## SCALE Codec
+
+Rust implementation of the SCALE data format for types used in the Polkadot SDK framework.
+
+The codec is implemented using the following traits:
+
+- [Encode](#encode)
+- [Decode](#decode)
+- [CompactAs](#compactas)
+- [HasCompact](#hascompact)
+- [EncodeLike](#encodelike)
+
+### Encode
+
+The [`Encode`](https://docs.rs/parity-scale-codec/latest/parity_scale_codec/trait.Encode.html){target=\_blank} trait handles data encoding into SCALE format and includes the following key functions:
+
+- `size_hint(&self) -> usize`: Estimates the number of bytes required for encoding to prevent multiple memory allocations. This should be inexpensive and avoid complex operations. Optional if the size isn’t known.
+- `encode_to<T: Output>(&self, dest: &mut T)`: Encodes the data, appending it to a destination buffer.
+- `encode(&self) -> Vec<u8>`: Encodes the data and returns it as a byte vector.
+- `using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R`: Encodes the data and passes it to a closure, returning the result.
+- `encoded_size(&self) -> usize`: Calculates the encoded size. Should be used when the encoded data isn’t required.
```

Comment:
```suggestion
- **`size_hint(&self) -> usize`** - estimates the number of bytes required for encoding to prevent multiple memory allocations. This should be inexpensive and avoid complex operations. Optional if the size isn’t known
- **`encode_to<T: Output>(&self, dest: &mut T)`** - encodes the data, appending it to a destination buffer
- **`encode(&self) -> Vec<u8>`** - encodes the data and returns it as a byte vector
- **`using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R`** - encodes the data and passes it to a closure, returning the result
- **`encoded_size(&self) -> usize`** - calculates the encoded size. Should be used when the encoded data isn’t required
```